pkg/resourcemanager/controller/node/reconciler.go

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2023 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package node
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/gardener/gardener/pkg/resourcemanager/apis/config"
+)
+
+// Reconciler manages taints on new Node objects to block scheduling of user workload pods until all node critical
+// components are ready.
+type Reconciler struct {
+	Client   client.Client
+	Config   config.NodeControllerConfig
+	Recorder record.EventRecorder
+}
+
+// Reconcile checks if the critical components not ready taint can be removed from the Node object.
+func (r *Reconciler) Reconcile(reconcileCtx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := logf.FromContext(reconcileCtx)
+
+	ctx, cancel := context.WithTimeout(reconcileCtx, time.Minute)
+	defer cancel()
+
+	node := &corev1.Node{}
+	if err := r.Client.Get(ctx, req.NamespacedName, node); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	if !NodeHasCriticalComponentsNotReadyTaint(node) {
+		return reconcile.Result{}, nil
+	}
```

Comment:
Correct.
What do you mean? What else should it cover?

Reply:
No, it's fine - I just remembered that people were asking about such "predicates" in the reconciler while we already have similar predicates in the controller. Generally, it can be confusing because it looks duplicative. Maybe a comment here would be helpful?

Reply:
Got it, done :)